glass/rng_dispatcher.py

```diff
@@ -0,0 +1,132 @@
+        return ()
+    if isinstance(size, int):
+        return (size,)
+    return size
+
+
+class JAXGenerator:
+    """JAX random number generation as a NumPy generator."""
+
+    __slots__ = ("key", "lock")
+    key: PRNGKeyArray
+    lock: Lock
+
+    @classmethod
+    def from_key(cls, key: PRNGKeyArray) -> Self:
+        """Wrap a JAX random key."""
+        if not isinstance(key, ArrayLike) or not issubdtype(key.dtype, prng_key):
+            msg = "not a random key"
+            raise ValueError(msg)
+        rng = object.__new__(cls)
+        rng.key = key
+        rng.lock = Lock()
+        return rng
+
+    def __init__(self, seed: int | Array, *, impl: str | None = None) -> None:
+        """Create a wrapper instance with a new key."""
+        self.key = key(seed, impl=impl)
+        self.lock = Lock()
+
+    @property
+    def __key(self) -> Array:
+        """Return next key for sampling while updating internal state."""
+        with self.lock:
+            self.key, key = split(self.key)
+        return key
+
+    def split(self, size: Size = None) -> Array:
+        """Split random key."""
+        shape = _s(size)
+        with self.lock:
+            keys = split(self.key, 1 + math.prod(shape))
+            self.key = keys[0]
+        return keys[1:].reshape(shape)
+
+    def spawn(self, n_children: int) -> list[Self]:
+        """Create new independent child generators."""
+        with self.lock:
+            self.key, *keys = split(self.key, num=n_children + 1)
+        return list(map(self.from_key, keys))
+
+    def random(self, size: Size = None, dtype: Shaped[Array, ...] = float) -> Array:
+        """Return random floats in the half-open interval [0.0, 1.0)."""
+        return uniform(self.__key, _s(size), dtype)
+
+    def normal(
+        self,
+        loc: float,
+        scale: float,
+        size: Size = None,
+        dtype: Shaped[Array, ...] = float,
+    ) -> Array:
+        """Draw samples from a Normal distribution (mean=loc, stdev=scale)."""
+        return loc + scale * normal(self.__key, _s(size), dtype)
+
+    def poisson(
+        self, lam: float, size: Size = None, dtype: Shaped[Array, ...] = float
+    ) -> Array:
+        """Draw samples from a Poisson distribution."""
+        return poisson(self.__key, lam, size, dtype)
+
+    def standard_normal(
+        self, size: Size = None, dtype: Shaped[Array, ...] = float
+    ) -> Array:
+        """Draw samples from a standard Normal distribution (mean=0, stdev=1)."""
+        return normal(self.__key, _s(size), dtype)
+
+    def uniform(
+        self,
+        low: int = 0,
+        high: int = 1,
+        size: Size = None,
+        dtype: Shaped[Array, ...] = float,
+    ) -> Array:
+        """Draw samples from a Uniform distribution."""
+        return uniform(self.__key, _s(size), dtype, low, high)
+
+
+def rng(array: NDArray[Any] | Array) -> JAXGenerator | np.random.Generator:
```

Comment:
The dispatching mechanism requires to dispatch on the basis of something. I looked around, and every function accessing an RNG accepts 1 or more arrays as arguments. So, this looks like the best solution to me.

The other solution is to check if a user has jax imported / installed and dispatch on the basis of that.